File: floors/erlang/src/office.erl
%%%-------------------------------------------------------------------
%%% @doc
%%% Office definitions for Floor 23
%%% @end
%%%-------------------------------------------------------------------
-module(office).

-export([all_offices/0, get_office/1]).

all_offices() ->
    [
        #{
            type => architecture,
            name => <<"Architecture Office">>,
            description => <<"Telecom-grade system design and OTP architecture">>
        },
        #{
            type => implementation,
            name => <<"Implementation Office">>,
            description => <<"gen_server and supervisor implementation">>
        },
        #{
            type => review,
            name => <<"Review Office">>,
            description => <<"Code review with telecom patterns verification">>
        },
        #{
            type => test,
            name => <<"Test Office">>,
            description => <<"EUnit and Common Test testing">>
        },
        #{
            type => security,
            name => <<"Security Office">>,
            description => <<"Process isolation and nine-nines reliability audit">>
        },
        #{
            type => manager,
            name => <<"Manager Office">>,
            description => <<"Project coordination and rebar3 management">>
        }
    ].

get_office(Type) ->
    lists:keyfind(Type, 1, all_offices()).

File: floors/haskell/Office.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}

module Office
    ( Office(..)
    , OfficeType(..)
    , allOffices
    , getOfficeDescription
    ) where

import Data.Aeson (ToJSON(..), object, (.=))
import Data.Text (Text)
import GHC.Generics (Generic)

data OfficeType
    = ArchitectureOffice
    | ImplementationOffice
    | ReviewOffice
    | TestOffice
    | SecurityOffice
    | ManagerOffice
    deriving (Show, Eq, Enum, Bounded, Generic)

data Office = Office
    { officeType :: OfficeType
    , officeName :: Text
    , description :: Text
    } deriving (Show, Eq, Generic)

instance ToJSON Office where
    toJSON office = object
        [ "type" .= show (officeType office)
        , "name" .= officeName office
        , "description" .= description office
        ]

allOffices :: [Office]
allOffices =
    [ Office ArchitectureOffice "Architecture Office" "Type-safe design and system architecture"
    , Office ImplementationOffice "Implementation Office" "Pure functional implementation with monadic effects"
    , Office ReviewOffice "Review Office" "Code review with property-based testing mindset"
    , Office TestOffice "Test Office" "QuickCheck and HUnit testing"
    , Office SecurityOffice "Security Office" "Type-level security and safe API design"
    , Office ManagerOffice "Manager Office" "Project coordination and dependency management"
    ]

getOfficeDescription :: OfficeType -> Text
getOfficeDescription ArchitectureOffice = "Type-safe design and system architecture"
getOfficeDescription ImplementationOffice = "Pure functional implementation with monadic effects"
getOfficeDescription ReviewOffice = "Code review with property-based testing mindset"
getOfficeDescription TestOffice = "QuickCheck and HUnit testing"
getOfficeDescription SecurityOffice = "Type-level security and safe API design"
getOfficeDescription ManagerOffice = "Project coordination and dependency management"
